### DISCUSSION TITLE
Allow uppercase characters for Subscript

### DISCUSSION THREAD

**Discussion by smdesai** [EvgSkv/logica#197] [General]
Would it be possible to allow uppercase characters for subscripts? As part of a query, I'm accessing JSON content which contains uppercase characters for a key.

**Comment by EvgSkv** [comment]
The rationale makes sense.
However this would be a significant change to the feel of the syntax, so I'm worried about making this change quickly.

Meanwhile, if you use SQLite you can access capital fields with `JsonExtract` command:
```
@Engine("sqlite");
T({A:1});
T({A: 2});
Q(a) :-
  T(x), a == JsonExtract(x, "$.A");
```

If you use some other engine, let me know, there is most likely a workaround for now too.

**Reply by smdesai** [reply]
Thanks for the explanation as it's only a suggestion that would make the syntax easier to read. However, I understand the concerns about making the change quickly. I use both SQLite and more recently Trino so for Trino it would be something like "JsonExtractScalar(payload, item);"

On a similar note, I notice in SQLite that lower case subscripts can be accessed using the dot notation. So the above example can be written as:

@Engine("sqlite");
T({a:1});
T({a: 2});
Q(a) :-
  T(x), a == x.a;

In Trino, this doesn't work but by changing the following method in "class Trino(Dialect):" to the following, the same behavior works fine. 

```
def Subscript(self, record, subscript):
  return "JSON_EXTRACT_SCALAR(%s, '$.%s')" % (record, subscript)
```

I don't see this change breaking anything, is that a correct assumption.

**Reply by EvgSkv** [reply]
Okay, it's subtle :-)

I just checked and it turns out exampled that you gave doesn't work in Trino, but as far as I can see for a different reason. Trino's syntax that Logica uses breaks if structure has the only element. The following example works for me:

```
@Engine("trino");
@With(T);
T({a:1,b:2});
T({a:2,b:3});
Q(x) :- T(y), x == y.a;
```
A run of `Q` result in:
```
 col0 
------
    2 
    1 
(2 rows)
```

Subscript in Logica is translated into subscript in SQL when Trino dialect is used. So the example above translates to:
```sql
WITH t_0_T AS (SELECT * FROM (
  
    SELECT
      (SELECT 1 AS a, 2 AS b) AS col0
   UNION ALL
  
    SELECT
      (SELECT 2 AS a, 3 AS b) AS col0
  
) AS UNUSED_TABLE_NAME  )
SELECT
  T.col0.a AS col0
FROM
  t_0_T AS T;
```

SQLite does not have native structs but has great JSON support. So Logica uses JSON in SQLite in the role of a struct.
Trino has structs, so the subscript is reserved for those.

Let me know if you have further comments or thoughts.

**Reply by smdesai** [reply]
I believe the problem I have is with different data types and I may be making a false assumption on how to access it. Thanks for the clarification.

**Reply by EvgSkv** [reply]
You are very welcome!

If you have further questions, including how to best handle your specific case, please don't hesitate to reach out.
